### PR TITLE
lm-eval compatibility

### PR DESCRIPTION
**Pull request by mmoffatt2**
This PR adds support for running the [LM-Eval](https://github.com/EleutherAI/lm-evaluation-harness) evaluation harness directly from our sample.py script. 

To support nanoGPT models with lm-eval, we needed to implement three functions (as defined [here](https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/model_guide.md)):

**generate_until**:
- Each request contains Instance.args : Tuple[str, dict] containing 1. an input string to the LM and 2. a dictionary of keyword arguments used to control generation parameters.
- Samples text until one of the stop words is reached.

**loglikelihood**:
- Each request contains Instance.args : Tuple[str, str] containing 1. an input string to the LM and 2. a target string on which the loglikelihood of the LM producing this target, conditioned on the input, will be returned.
- Computes the log probability of generating the target string conditioned on the input.

**loglikelihood_rolling**:
- Each request contains Instance.args : Tuple[str], which is an input string to the model whose entire loglikelihood, conditioned on purely the EOT token, will be calculated.
- Computes the loglikelihood of producing each piece of text given no starting input.

**Example Invocation**:
I trained a small model for testing purposes:
`python3 train.py --dataset="wikitext103"`
`python3 sample.py --out_dir out --lm_eval_tasks arc_easy,hellaswag`

**Example Output**:
image

**sample.py Arguments**:
--lm_eval_tasks: comma-separated list of tasks for evaluation (full list of tasks and their descriptions can be found [here](https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/tasks/README.md))
--batch_size: batch size for LM-Eval
--lm_eval_results_output: optional output path for the JSON results

**Requires**:
lm-eval
accelerate==0.32.0
peft==0.10.0

### CODE REVIEW

**Comment by klei22** [comment]
@mmoffatt2 Instead of adding to each of the CI github workflows, instead just add these new dependencies to the `requirements_cpu.txt` and to the `requirements_gpu.txt`.

File: sample.py
Comment (gkielian) [inline sample.py:1]: @mmoffatt2 the signature for generate with the `stop_string` in model.py was changed to take instead a list `stop_strings`.

I think to ensure compatibility we need to change the "stop_string" in sample.py to "stop_strings" and taking in a list of possible stop strings.

File: sample.py [hunk 1]
```diff
@@ -499,6 +515,46 @@ def main():
     if args.rope_length:
         model.update_rope_length(args.rope_length)
 
+    if args.lm_eval_tasks:
```
Comment (gkielian) [inline sample.py:518]: Instead of running once, convert this to a function, and also have this in the gpt_lm_eval_wrapper.py.

We basically just want to have a single function that we import to sample.py, as it's a major file and we should leave implementation details to files we're doing the importing.

I'm hoping we can eventually call the function from train.py too, which could allow us to include this in tensorboard logs, and call at intervals like the max_sample_tokens or validation loss estimations.

**Review by gkielian — CHANGES_REQUESTED** [review]
Have two feedbacks:

1. Need to update stop string in sample.py to ensure it is compatible with the new "stop_strings" function signature modified into model.py
2. Make the lmeval a single function call, the function should live in the gpt_lm_eval_wrapper.py.

File: sample.py (hunk 1 above)
Comment (mmoffatt2) [reply sample.py:518]: I created two functions: one for creating the NanoGPTLM model and one for evaluating and saving. Thus, if we want to reuse the code in train.py, we can instantiate the model once, not on every function call.

Comment (gkielian) [reply sample.py:518]: Looks good!

I'll merge this in now, and please begin looking into folding into the train.py as well

**Review by gkielian — APPROVED** [review]
Looks good, and please begin looking into incorporating into train.py
